lib/watir/locators/element/locator.rb

```diff
@@ -70,39 +71,58 @@ def using_selenium(filter = :first)
 
         def using_watir(filter = :first)
           selector = selector_builder.normalized_selector
-          visible = selector.delete(:visible)
-          visible_text = selector.delete(:visible_text)
           tag_name = selector[:tag_name].is_a?(::Symbol) ? selector[:tag_name].to_s : selector[:tag_name]
           validation_required = (selector.key?(:css) || selector.key?(:xpath)) && tag_name
 
           if selector.key?(:index) && filter == :all
             raise ArgumentError, "can't locate all elements by :index"
           end
-          idx = selector.delete(:index) unless selector[:adjacent]
+
+          filter_selector = delete_filters_from(selector)
+          filter_selector[:tag_name] = tag_name if validation_required
+
+          query_scope = ensure_scope_context
+
+          if filter_selector.key?(:label) && selector_builder.should_use_label_element?
+            label = label_from_text(filter_selector.delete(:label)) || return
```

Comment:
Does `|| return` still belong here now that it isn't in `#wd_find_by_regexp_selector`?

Also, what do we think about breaking up this method more?
Can we pull these lines out into something like:
```
query_scope = set_query_scope(selector, filter_selector)
```

Reply:
`|| return` is still required. If the label cannot be found, there is no point in continuing - ie return element not found.

I had debated breaking up the code. I still think a lot of this should be the selector builders responsibility. However, given I don't know when I'll get around to tackling that, I'll at least look at tidying up the current code.

Reply:
Ah, I see it now. Thanks

Reply:
How do you see the responsibilities needing to be split? The only thing that seems obvious to me to figure out how to push it into the SelectorBuilder class is the predicate generation. Everything else is some kind of filtering or validation.

Reply:
I feel like deciding what should be in the XPath vs a filter would be part of the selector builder. The selector builder would:
* Be given all of the properties (selectors plus filters)
* Take which ever properties it can handle and build the XPath
* Return the XPath along with a list of properties that it didn't know what to do with (ie would eventually be the filters)

I guess I see the Locator class as managing the workflow. It shouldn't be deciding what the selector builder can or cannot do.